## this test is required when 2 categorical features( from same population) are to be compared 

### it is to determine whether there is a significant association between the two variables.

In [1]:
import pandas as pd
import scipy.stats as stats
import numpy as np
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
dataset = sns.load_dataset('tips')

In [3]:
dataset.describe(include='all')

total_bill         tip   sex smoker  day    time        size
count   244.000000  244.000000   244    244  244     244  244.000000
unique         NaN         NaN     2      2    4       2         NaN
top            NaN         NaN  Male     No  Sat  Dinner         NaN
freq           NaN         NaN   157    151   87     176         NaN
mean     19.785943    2.998279   NaN    NaN  NaN     NaN    2.569672
std       8.902412    1.383638   NaN    NaN  NaN     NaN    0.951100
min       3.070000    1.000000   NaN    NaN  NaN     NaN    1.000000
25%      13.347500    2.000000   NaN    NaN  NaN     NaN    2.000000
50%      17.795000    2.900000   NaN    NaN  NaN     NaN    2.000000
75%      24.127500    3.562500   NaN    NaN  NaN     NaN    3.000000
max      50.810000   10.000000   NaN    NaN  NaN     NaN    6.000000

In [4]:
dataset.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

#### to find the relationship(counts) between two categorical variables. crosstab is very efficient

## null hypothesis = for population , sex and smoker are not related.
## alternate = sex and smoker features are related.

In [5]:
observed_values = pd.crosstab(dataset.sex,dataset.smoker).values

In [6]:
print('Observed values - \n', observed_values)

Observed values - 
 [[60 97]
 [33 54]]


In [98]:
# return values are (ch2, pvalue, dof, expected_values)
val = stats.chi2_contingency(observed_values)

In [8]:
val

(0.008763290531773594, 0.925417020494423, 1, array([[59.84016393, 97.15983607],
        [33.15983607, 53.84016393]]))

## as p value= 0.925 is more than 0.05 , null hypothesis cannot be not rejected. there is no relationship between features. 

---
## take another example where both categories are related.

In [140]:
classA = [['Yes','No'][np.random.randint(2)] for i in range(20)]

In [142]:
classB = []
for i in classA:
    if i == 'No':
        classB.append('Black')
    else:
        classB.append('White')
        

In [143]:
print(classA)
print(classB)
# BOth are exactly same list. Lets see chi square pick it or not.

['No', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'Yes', 'Yes', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No']
['Black', 'Black', 'White', 'White', 'White', 'White', 'White', 'White', 'White', 'White', 'Black', 'White', 'White', 'Black', 'White', 'Black', 'Black', 'Black', 'Black', 'Black']


In [144]:
observed_values =pd.crosstab(np.array(classA),np.array(classB)).values

In [145]:
observed_values

array([[ 9,  0],
       [ 0, 11]], dtype=int64)

In [147]:
val = stats.chi2_contingency(observed_values)
# return values are ch2, pvalue, dof, expected_values


In [148]:
val

(16.163656769717377, 5.8098162290722885e-05, 1, array([[4.05, 4.95],
        [4.95, 6.05]]))

## Pvalue is 5.8e-5, so obviously it is less than 0.05, null hypothesis is rejected. both features are related

---
## lets update the lists and see the results

### changing multiple of 5 places.

In [149]:

classC = classB
for i,x in enumerate(classC):
    if i %5 == 0:
        if x == 'Black':
            classC[i]= 'White'
        else:
            classC[i] = 'Black'
            

In [150]:
print(classA)
print(classC)

['No', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'Yes', 'Yes', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No']
['White', 'Black', 'White', 'White', 'White', 'Black', 'White', 'White', 'White', 'White', 'White', 'White', 'White', 'Black', 'White', 'White', 'Black', 'Black', 'Black', 'Black']


In [151]:
observed_values1 = pd.crosstab(np.array(classA),np.array(classC)).values

In [152]:
observed_values1

array([[ 6,  3],
       [ 1, 10]], dtype=int64)

In [153]:
val1 = stats.chi2_contingency(observed_values1)

In [154]:
val1

(4.903984903984904, 0.0267947959446109, 1, array([[3.15, 5.85],
        [3.85, 7.15]]))

## Pvalue is 0.026, so obviously it is less than 0.05, null hypothesis can be rejected. There are 95% chances that both features are related


---
## another scenario

multiple of 3

In [155]:
classD = classB
for i,x in enumerate(classD):
    if i %3 == 0:
        if x == 'Black':
            classD[i]= 'White'
        else:
            classD[i] = 'Black'

In [156]:
observed_values2 = pd.crosstab(np.array(classB),np.array(classE)).values

In [157]:
observed_values2

array([[4, 8],
       [5, 3]], dtype=int64)

In [158]:
val2 = stats.chi2_contingency(observed_values2)

In [159]:
val2

(0.6818181818181821, 0.4089613420368744, 1, array([[5.4, 6.6],
        [3.6, 4.4]]))

## Pvalue is 0.4089, so obviously it is more than 0.05, null hypothesis cannot be rejected. that both features are not related